broker_test.go

```diff
@@ -52,36 +55,45 @@ func TestBrokerAccessors(t *testing.T) {
 }
 
 func TestSimpleBrokerCommunication(t *testing.T) {
-	mb := NewMockBroker(t, 0)
-	defer mb.Close()
-
-	broker := NewBroker(mb.Addr())
-	conf := NewConfig()
-	conf.Version = V0_10_0_0
-	err := broker.Open(conf)
-	if err != nil {
-		t.Fatal(err)
-	}
-
 	for _, tt := range brokerTestTable {
+		Logger.Printf("Testing broker communication for %s", tt.name)
+		mb := NewMockBroker(t, 0)
 		mb.Returns(&mockEncoder{tt.response})
-	}
-	for _, tt := range brokerTestTable {
+		broker := NewBroker(mb.Addr())
+		// Set the broker id in order to validate local broker metrics
+		broker.id = 0
+		conf := NewConfig()
+		conf.Version = V0_10_0_0
+		// Use a new registry every time to prevent side effect caused by the global one
+		conf.MetricRegistry = metrics.NewRegistry()
+		err := broker.Open(conf)
+		if err != nil {
+			t.Fatal(err)
+		}
 		tt.runner(t, broker)
+		err = broker.Close()
+		if err != nil {
+			t.Error(err)
+		}
+		// Wait up to 500 ms for the remote broker to process requests
+		// in order to have consistent metrics
+		if err := mb.WaitForExpectations(500 * time.Millisecond); err != nil {
```

Comment:
I use a dedicated `MockBroker` for each runner otherwise the statistics to validate metrics against would contains data from previous runs.

But I found a race condition that often manifests itself for unacknowledged produce request (i.e. `RequiredAcks` set to `NoResponse`) but can happen for standard request too when using a `MockBroker` for each runner.
That is the runner would return early because the produce request has been written to the socket and invoking `MockBroker.Close()` gets invoked before `MockBroker` is able to process that request (the request might not even have been read by `MockBroker` but already acknowledged by the TCP stack) and therefore consume from the expectations channel leading to an error about expectation not being satisfied or that the statistics about the bytes written by the `MockBroker` being 0.

On a regular produce request, the runner would block on reading the response and by then it is very likely that the `MockBroker` has consumed his expectations as well as update his statistics but there is a very tight window where the runner would be reading the statistics to validate the metrics before the `MockBroker` updates them.

My first fix was to add a magic 100 ms sleep for the `ProduceRequest (NoResponse)` but like every magic sleep that was futile and then I believe I ran into the second case once probably because the integration server being loaded.
The working fix I came with consists of adding that `WaitForExpectations` method to be sure that we received and processed the expected requests including updating statistics.
I added a timeout of 500ms to be sure that it would not blocked on failing test or very loaded integration server, it generally finishes under a couple of ms.

I guess it would be good to document those scenarios better on both `broker_test.go` and `mockbroker.go`.